jujugui/static/gui/src/test/test_env_api.js

```diff
@@ -1298,6 +1299,222 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
+        // Perform the request.
+        const args = {
+          applicationName: 'wordpress',
+          charmURL: '',
+          channel: 'candidate',
+          resources: [{Name: 'myres', File: 'myres.tar.bz2'}]
+        };
+        env.addPendingResources(args, (err, ids) => {
+          assert.equal(err, 'invalid arguments: ' + JSON.stringify(args));
+          assert.deepEqual(ids, {});
+          done();
+        });
+      });
+
+      it('fails when no resources are provided', done => {
+        // Perform the request.
+        const args = {
+          applicationName: 'wordpress',
+          charmURL: 'wordpress-42',
+          resources: []
+        };
+        env.addPendingResources(args, (err, ids) => {
+          assert.equal(err, 'invalid arguments: ' + JSON.stringify(args));
+          assert.deepEqual(ids, {});
+          done();
+        });
+      });
+
+      it('fails when invalid resources are provided', done => {
+        // Perform the request.
+        env.addPendingResources({
+          applicationName: 'wordpress',
+          charmURL: 'wordpress-42',
+          resources: [{File: 'bad'}]
+        }, (err, ids) => {
+          assert.equal(err, 'resource without name: [{"File":"bad"}]');
+          assert.deepEqual(ids, {});
+          done();
+        });
+      });
+
+      it('handles global failures while adding resources', done => {
+        // Perform the request.
+        env.addPendingResources({
+          applicationName: 'redis',
+          charmURL: 'redis-0',
+          resources: [{Name: 'red', File: 'red.tgz'}]
+        }, (err, ids) => {
+          assert.strictEqual(err, 'bad wolf');
+          assert.deepEqual(ids, {});
+          done();
+        });
+        // Mimic response.
+        conn.msg({'request-id': 1, error: 'bad wolf'});
+      });
+
+      it('handles local failures while adding resources', done => {
+        // Perform the request.
+        env.addPendingResources({
+          applicationName: 'redis',
+          charmURL: 'redis-0',
+          resources: [{Name: 'red', File: 'red.tgz'}]
+        }, (err, ids) => {
+          assert.strictEqual(err, 'bad wolf');
+          assert.deepEqual(ids, {});
+          done();
+        });
+        // Mimic response.
+        conn.msg({
+          'request-id': 1,
+          response: {error: {message: 'bad wolf'}}
+        });
+      });
+
+      it('handles unexpected failures while adding resources', done => {
+        // Perform the request.
+        env.addPendingResources({
+          applicationName: 'redis',
+          charmURL: 'redis-0',
+          resources: [{Name: 'red', File: 'red.tgz'}]
+        }, (err, ids) => {
+          assert.strictEqual(err, 'unexpected response: {"pending-ids":[]}');
+          assert.deepEqual(ids, {});
+          done();
+        });
+        // Mimic response.
+        conn.msg({'request-id': 1, response: {'pending-ids': []}});
+      });
+    });
```

Comment:
This is strictly for me, but shouldn't all the tests have this conn.msg part? I'm not sure what's triggering the cb properly in cases w/o it.

Again, I'm asking mostly b/c it's been long enough since I last dealt with a lot of js code that I think I've just been confused by the cb chain, not that I actually think there's something wrong here.